### ISSUE TITLE
Openidconnect 1.0.0

### ISSUE THREAD

**Issue by HanaGemela** [owncloud/openidconnect#66]
### Product approval

- [x] Verify transifex commits are in master: If a folder https://github.com/owncloud/<APPNAME>/tree/master/l10n exists, then https://drone.owncloud.com/owncloud/<APPNAME> should receive nightly '[tx] updated from transifex` ... - no commits in last two months but no user facing strings -> OK
- [x] Ping pmaier to check `info.xml` description and screenshot
- [x] Get approval on translations from pmaier

### QA

- [x] Generate Changelog
- [x] Create **QA** Milestone and set the description to $version
- [x] Move items from `development` to `qa`
- [x] Create release branch with the format "release-$version"
- [ ] Create "$version Testplan"

### Documentation

- [x] Changes in the docs needed? If so create issue in owncloud/docs. https://github.com/owncloud/docs/pull/2447
- [ ] https://github.com/owncloud/docs/pull/2830

### Marketing

- [x] Ping @owncloud/marketing

### Build

All actions to be done on the release branch from here:
- [x] Bump version in `info.xml` (no version suffix)
- [x] Check php version dependency in `info.xml`
- [x] Check minimum ownCloud version dependency in `info.xml`
- [x] Change CHANGELOG.md 'Unreleased' to '$version' (today's date)

#### Beta/RC

- [x] Create github tag for `v${version}RC1` on release branch
- [x] Build and sign RC tarball from github tag `v${version}RC1` (see
  handbook for how to build)
- [x] Upload as Github release using changelog entries
- [x] Test and document in testplan => ~~Final "Go" from QA (set 'QA-ready' label)~~ => else do another RC after fixing issues in the "release-$version" branch
####
- [x] Create github tag for `v${version}RC5` on release branch
- [x] Build and sign RC tarball from github tag `v${version}RC5` (see
  handbook for how to build)
- [x] Upload as Github release using changelog entries
- [ ] Test and document in testplan => Final "Go" from QA (set 'QA-ready' label) => else do another RC after fixing issues in the "release-$version" branch

#### Final

- [x] Check that no documentation issues are pending https://github.com/owncloud/docs/pull/2447
- [x] Rename last RC tarball to final
- [x] Create github tag for final `v${version}` on release branch
- [x] Upload as Github release using changelog entries

### Publishing

- [x] Upload tarball to marketplace (see handbook for details)
- [x] Publish tarball for customers (see handbook for details) - https://customer.owncloud.com and https://portal.owncloud.com
- [x] Send mail to release-coordination with link to this ticket and to the changelog
- [x] Copy email to rocket chat `#updates` channel

### Post-release

- [x] Rename QA milestone to `$version`
- [x] Add a new 'Unreleased' section in CHANGELOG.md
- [x] Create PR to merge changes back to the `master` branch with
  description "Closes #XYZ" with the release ticket number
- [x] Close the milestone
- [x] Ensure [release template](https://github.com/owncloud/handbook/edit/master/modules/ROOT/examples/releases/server_app_release_template.md) is up to date

- [x] Add the app to the normal server tarball definitions

## Issues found during QA
1.0.0RC4:

- [x] (1) Android bad behaviours depending on the version https://github.com/owncloud/openidconnect/issues/66#issuecomment-694718837  
- [x] (2) iOS does not redirect to the callback URL after finishing auth process https://github.com/owncloud/openidconnect/issues/66#issuecomment-694201601
- [ ] (3) Desktop behaviour  and Mobile Clients feedback https://github.com/owncloud/openidconnect/issues/66#issuecomment-694087774

- [ ] Find a solution for "authorization successful" page @IljaN https://jira.owncloud.com/browse/OCIS-723
- [ ] Find a solution for "switch user" page @IljaN https://jira.owncloud.com/browse/OCIS-722

**Comment by mmattel** [comment]
@micbar @pmaier1 what is the plan to release OIDC ?

**Comment by michaelstingl** [comment]
> @micbar @pmaier1 what is the plan to release OIDC ?

no ETA yet

**Comment by pmaier1** [comment]
- Test setup and log in / log out with Kopano Konnect
- Test OAuth2 => OIDC migration with clients (running both oC 10 apps in parallel)

**Comment by HanaGemela** [comment]
The app won't build cc @phil-davis 
`make: *** No rule to make target 'dist'.  Stop.`

**Comment by phil-davis** [comment]
How was it built last time?
(0.2.0 or 1.0.0RC1)

I am not familiar with why this app have "make"  code split between `Makefile` and `Makefile.release`

Having apps with different structure of "make"  files means that developers, release builders, and... have to somehow know specific build commands for each app. That is not good.

**Comment by mmattel** [comment]
As I see progress here to build a RC (and maybe later a final first release),
I want to make you aware, that docs is not ready for it, see https://github.com/owncloud/docs/pull/2447

**Comment by mmattel** [comment]
@HanaGemela @pmaier1 pls see https://github.com/owncloud/openidconnect/pull/87 which is open and maybe affects a successful first release

**Comment by HanaGemela** [comment]
Merge the release branch back to include more fixes 
Merge two makefiles into one

**Comment by HanaGemela** [comment]
@davitol ready for testing https://github.com/owncloud/openidconnect/releases/tag/v1.0.0RC2

**Comment by HanaGemela** [comment]
Test with only oC 10.5 @davitol 
Include testing with oAuth2

**Comment by davitol** [comment]
> Test with only oC 10.5 @davitol

Tested with oC 10.5 + Openidconnect 1.0.0.RC2 + https://github.com/panva/node-oidc-provider/

- [ ] regular user
- [ ] LDAP user
- [ ] oAuth2 enabled

**Comment by davitol** [comment]
> Include testing with oAuth2

@HanaGemela Should it be tested against oauth2 0.4.4 (unreleased) or 0.4.3.?

**Comment by michaelstingl** [comment]
Please also cover migration from pre-OIDC-clients (desktop and mobile) to the latest ones with OIDC support.

**Comment by davitol** [comment]
We should not use the IDP described in the README section https://github.com/owncloud/openidconnect  It is only for DEV purpose.

Instead of it, we should use Kopano https://github.com/owncloud-docker/compose-playground/pull/4

**Comment by jnweiger** [comment]
My current state: https://github.com/owncloud-docker/compose-playground/pull/51

**Comment by pmaier1** [comment]
To be decided: Known issue or release blocker https://github.com/owncloud/enterprise/issues/4191 ?

**Comment by pmaier1** [comment]
This should be part of 1.0.0: https://github.com/owncloud/openidconnect/pull/98

**Comment by jnweiger** [comment]
Rc3 has #98

**Comment by jnweiger** [comment]
Test setup deployed via hetzner_deploy/oake_openidconnect_test.sh

**Comment by pmaier1** [comment]
Added: This should be part of 1.0.0 https://github.com/owncloud/openidconnect/pull/100

**Comment by jnweiger** [comment]
@davitol test setup deployed at https://owncloud.oidc-jw-qa.owncloud.works/

**Comment by jnweiger** [comment]
>     * regular user

@pmaier1  expects this to fail in this setup. Please try anyway. :-)

**Comment by davitol** [comment]
Tested with oC 10.5 + Openidconnect 1.0.0.RC4 + Firefox WebBrowser 

Scenarios tested:

- [x] regular user
- [x] LDAP user
- [x] oAuth2 0.4.3 enabled
- [ ] oAuth2 0.4.4 RC enabled
- [x] Test with iOS app 11.4. **Note** Login page shows a warning info related to oauth2 even when oauth2 is not enabled (see screenshot below)
- ❌  Test with Android app  Browser does not redirect back to the Android app
- ❌  Test with Desktop Client (Mac 2.7.0.) Browser does not redirect back to the Desktop Client. It stays in 'Please switch to your browser to proceed.


![ioSAPP](https://user-images.githubusercontent.com/10596560/93446964-2113fc80-f8d2-11ea-8f5e-20465da4dc6a.jpeg)

**Comment by jnweiger** [comment]
@michaelstingl @pmaier1 see above test plan an results.

**Comment by jnweiger** [comment]
>     * Test OAuth2 => OIDC migration with clients (running both oC 10 apps in parallel)

@patrickjahns do you have details what the migration means exactly?

**Comment by jesmrec** [comment]
> @davitol test setup deployed at https://owncloud.oidc-jw-qa.owncloud.works/

Tests with mobile devices over such instance

- Android (2.15.2 version)

after entering credentials, an error is returned

![Screenshot 2020-09-17 at 13 26 03](https://user-images.githubusercontent.com/14894746/93464388-6263d680-f8e9-11ea-96cf-673441481011.png)

Checking the request flow, i notice that the `authorize` endpoint return a 302 with a redirection to error with a description

`redirect?error=login_required&error_description=IdentifierIdentityManager%3A%20not%20signed%20in&state=31a64353c16f970286788d411dd8e697`

which next request causes a `503`. Something bad happens in authorization stage with such client id.

- iOS (11.4 version)

Login process works fine, but at the end the browser does not redirect to the app and stays in the browser

![Sep-17-2020 14-21-40](https://user-images.githubusercontent.com/14894746/93470030-980cbd80-f8f1-11ea-8464-e029f5a0a8a5.gif)

my suspicion:

the request with the `authorization_code`, `scope` and `state` :

`/apps/openidconnect/redirect?code=...&scope=openid%20profile%20email&session_state=...&state=... HTTP/2.0`

returns 303 with 

`location:  https://owncloud.oidc-jw-qa.owncloud.works/apps/files/   `

such location is a web site to be opened with the browser. OAuth2 flow returns in that header the correct redirection path:  `oc://ios.owncloud.com?code=...` that is the callback to the app. 

The app sends in advance the correct redirection in the authorization request, before letting the control to the browser:

```
response_type: code
redirect_uri:  oc://ios.owncloud.com
client_id:     mxd5OQDk6es5LzOzRvidJNfXLUZS2oN3oUFeXPP8LpPrhx3UroJFduGEYIBOxkY1
```

somewhere, the redirection is missing.

If further info is needed, please ping me.

**Comment by davitol** [comment]
>     * Android (2.15.2 version)
> 
> 
> after entering credentials, an error is returned

It does not happen to me using the same Android version than you @jesmrec .  Login process works fine, but at the end the browser does not redirect to the app and stays in the browser as it happens to you in iOS.

**Comment by jesmrec** [comment]
> > ```
> > * Android (2.15.2 version)
> > ```
> > 
> > 
> > after entering credentials, an error is returned
> 
> It does not happen to me using the same Android version than you @jesmrec . Login process works fine, but at the end the browser does not redirect to the app and stays in the browser as it happens to you in iOS. (Also I get in Android the same oAuth2 warning)

Something strange... i have spread the tests to more devices:

- Android 11 device: Same behaviour as you: list of files is shown in the web browser
- Android 9 device: Same as Android 11 and yours
- Android 7 device: Behaviour mentioned [here](https://github.com/owncloud/openidconnect/issues/66#issuecomment-694201601). 

As mentioned difference is in the header `location` in the response of the `signin/v1/identifier/_/authorize` request:

Android 9 & 11: `https://owncloud.oidc-jw-qa.owncloud.works/index.php/apps/openidconnect/redirect?code=...&scope=profile%20email%20openid&session_state=...&state=...`  ✅ 

Android 7: ` https://owncloud.oidc-jw-qa.owncloud.works/index.php/apps/openidconnect/redirect?error=login_required&error_description=IdentifierIdentityManager%3A%20not%20signed%20in&state=...` ❌   

All of them with the 2.15.2 Android release `b089032` , performing the same changes. Any further idea??

**Comment by pmaier1** [comment]
> @pmaier1 do you have details what the migration means exactly?

In current customer scenarios all clients (desktop / iOS / Android) are authenticated against the server using either basic auth or OAuth2.

Generally speaking, OIDC is an addition to the OAuth2 standard. For client authorization, we have to consider them as two different, mutually exclusive ways of doing it. Apart from that, OIDC shall supersede OAuth2 at some point. That's why we need to test the migration.

Practically this means, we need to have an exemplary oC instance with OAuth2 and authorized clients and need to test a migration to OIDC. For this purpose, both apps can be enabled in parallel (requires `oauth2` 0.4.4!). 

An example scenario would be:
1. In the exemplary instance, have a desktop client authorized with OAuth2
2. Enable the fully configured `openidconnect` app in parallel
3. Check how the desktop client behaves (especially when tokens are refreshed)
4. Understand what's necessary to move the desktop client from OAuth2 authorization to OIDC

Hope this helps. Please add all scenarios around this for all clients to the testplan.

**Comment by michaelstingl** [comment]
> Please also cover migration from pre-OIDC-clients (desktop and mobile) to the latest ones with OIDC support.

Following scenarios come to my mind:

### Desktop

**Client version/feature change**

- Desktop 2.6 connected to oC 10.5 with OIDC + OAuth ==> Desktop 2.7 connected to oC 10.5 with OIDC
- Desktop 2.6 connected to oC 10.5 with OIDC + OAuth ==> Desktop 2.7 connected to oC 10.5 with OIDC + OAuth

**Server feature change**

- Desktop 2.7 connected to oC 10.5 with OAuth ==> Desktop 2.7 connected to oC 10.5 with OIDC
- Desktop 2.7 connected to oC 10.5 with OAuth ==> Desktop 2.7 connected to oC 10.5 with OIDC + OAuth

### Android

**Server feature change**

- Android 2.15 connected to oC 10.5 with OAuth ==> Android 2.15 connected to oC 10.5 with OIDC
- Android 2.15 connected to oC 10.5 with OAuth ==> Android 2.15 connected to oC 10.5 with OIDC + OAuth

### iOS

**Server feature change**

- iOS 11.4 connected to oC 10.5 with OAuth ==> iOS 11.4 connected to oC 10.5 with OIDC
- iOS 11.4 connected to oC 10.5 with OAuth ==> iOS 11.4 connected to oC 10.5 with OIDC + OAuth

**Comment by jesmrec** [comment]
```
Android
Server feature change

Android 2.15 connected to oC 10.5 with OAuth ==> Android 2.15 connected to oC 10.5 with OIDC
Android 2.15 connected to oC 10.5 with OAuth ==> Android 2.15 connected to oC 10.5 with OIDC + OAuth
```
does basic ==> OIDC make sense?

> @jesmrec question: which is the expected behaviour in migration? transparent for user? re-connection?

regarding this, together with the scenarios @michaelstingl has mentioned above:

`Android 2.15 connected to oC 10.5 with OAuth2==> Android 2.15 connected to oC 10.5 with OIDC`

OAuth2 and OIDC are managed by different apps, and they handle different tokens. In (mobile) clients, every account in stored internally with its confidential info (keychain, Android password manager), so token will be invalid  after migration, unless OAuth2 tokens are also migrated in backend DB (probably not too easy, more stuff and elaboration would be required). Reauthentication after migration to OIDC in (mobile) clients sounds the straightforward way.

`Android 2.15 connected to oC 10.5 with OAuth ==> Android 2.15 connected to oC 10.5 with OIDC + OAuth`

Does this mean, that both OIDC and OAuth2 are available at the time?  if both OIDC and OAuth2 are available, (mobile) clients would get OIDC for new connections, because they check `.well-known/opeinid-configuration` in advance if `Bearer` is present ... so, re-authentication also required to jump OAuth to OIDC. 

(everything applies to iOS as well)

btw, i uploaded a first (generical) testplan for OIDC, to complete with (maybe) more cases: 
https://github.com/owncloud/QA/blob/master/Mobile/GenericTPs/OIDC.md

Separately:

- User experience (check if OIDC is correct in terms of UI flow, what the user will see)
- Request flow: check if the internal stuff is correct, without checking the internal idP stuff deeply.
- Migrations

open to any suggestion

**Comment by michaelstingl** [comment]
Recycling of the OAuth2 token for the OIDC connection is not planned. Migration would require entering the password again I think.

> (mobile) clients would get OIDC for **new connections**, because they check …

How to migrate an **existing** account OAuth to OIDC? Can the server send 401 to the clients to trigger re-login with OIDC?

**Comment by jesmrec** [comment]
> How to migrate an existing account OAuth to OIDC? Can the server send 401 to the clients to trigger re-login with OIDC?

with "new connections" i meant new accounts to attach, not the existing ones before migration (sorry if it was confusing). As you mentioned, the jump will require re-authentication.

**Comment by michaelstingl** [comment]
Yeah, fully understood the behavior for new accounts. But how can a server admin trigger all connected clients to migrate from OAuth to OIDC? Did you found a way?

**Comment by jesmrec** [comment]
Not really. After migrating, all valid OAuth2 tokens should be invalidated and clients will detect an authentication failure, prompting users to enter credentials again (at least, mobile clients do that). Then, clients detect OIDC and so on. So, the server admin must assure that OAuth2 credentials (token) are no longer valid, but i think this is the expected behaviour in advance. We need to test in a real environment.

**Comment by jnweiger** [comment]
Seen with desktop testpilotcloud version 2.7.0daily20201004 (build 2209) on Ubuntu 18.04:
When oauth2 0.4.3 is enabled together with openidconnect-1.0.0-RC4 the desktop client complains with

![image](https://user-images.githubusercontent.com/1108546/95342906-35875b80-08b8-11eb-92ea-477730a9c53c.png)

Token confusion?

**Comment by jnweiger** [comment]
@jesmrec We've identified a misconfiguration in my test setup today. One line was missing in the caddyfile.
We need to re-evaluate https://github.com/owncloud/openidconnect/issues/66#issuecomment-694087774 et al.

**Comment by mmattel** [comment]
With all the tuning and setup made for OIDC, pls consider to share this knowledge in docs...

**Comment by jesmrec** [comment]
Latest checks with mobile against  `https://owncloud-1-0-0rc4.oidc-jw-qa.owncloud.works`

**- Android bad behaviours depending on versions (https://github.com/owncloud/openidconnect/issues/66#issuecomment-694718837)**

Fixed. Tested with Android 7, 9 and 11. All of them finish the authentication and authorization process in browser and then, redirection to the app is correct. 

**- iOS does not redirect to the app after finishing the process (https://github.com/owncloud/openidconnect/issues/66#issuecomment-694201601)**

Fixed. Redirection is done and control hands over to the app.

About the test plan, divided in three sections:

1) User experience (check if OIDC is correct in terms of UI flow, what the user will see) ✅ 

2) Request flow: check if the internal stuff is correct, without checking the internal idP stuff deeply. ✅ 

3) Migrations: i tried to test this by disabling the `openidconnect` app and enabling the `oauth2` app to "simulate" a change in the authentication methods over the same instance. It was not posible because server returns a correct response for the request against `.well-known/openid-configuration`, no matter which auth method is enabled.

**Comment by jnweiger** [comment]
https://github.com/owncloud/openidconnect/releases/tag/v1.0.0RC5

**Comment by jesmrec** [comment]
New checks with `https://oc-10-5-0.oidc-1-0-0rc5.jw-qa.owncloud.works/` . Thanks @jw for building the environment!

Different steps in the migration, tested one by one:

**1) OAuth2 installed and enabled (0.4.4). Some tests performed, since this is also an RC**

|  - | Check URL | Select auth method (OAuth2) | Authorization | Token renewal | Token revokal |
| :-: | :-: | :-: | :-: | :-: | :-: |
| **Android** | ✅  |✅  |✅  |✅  |✅  |
| **iOS** |  ✅  |✅  |✅  |✅  |✅  |

From my side (mobile clients), OAuth 0.4.4RC1 works fine. If some specific scenario needs more attention, let me know.

**2) Make `.well-known/openid-configuration` visible**

At this point, OpenId Connect is still disabled, but the config file is already visible. So, it is only checked the file visibility:

- Android: gets 200 in the `.well-known/openid-configuration` request ✅  
- iOS: gets 200 in the `.well-known/openid-configuration` request ✅  

**3) Enable OpenId Connect (OAuth2 is still enabled)**

In first instance, both Android and iOS continue working with the OAuth2 token granted in 1). No interruption for the user.
When token expires, everything goes fine because OAuth2 app is still enabled.

At this point, only two ways to switching to OpenID Connect in the mobile clients:

- Use the `Edit` option to re- authenticate
- Delete and create again

**4) Disable OAuth2 (OpenID connect enabled)**

- Android: App crashes. ❌ 

1. Regular request is sent -> returned 401 as unauthorized
2. Sent request to `oauth2/api/v1/token ` -> returns 302 pointing to `https://oc-10-5-0.oidc-1-0-0rc5.jw-qa.owncloud.works/login`and an exception happens:

```
2020-10-16 13:17:11.470 27944-27968/? E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #1
    Process: com.owncloud.android, PID: 27944
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$4.done(AsyncTask.java:415)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:305)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:923)
     Caused by: java.lang.NullPointerException
        at java.io.Reader.<init>(Reader.java:78)
        at java.io.InputStreamReader.<init>(InputStreamReader.java:72)
        at net.openid.appauth.Utils.readInputStream(Utils.java:36)
        at net.openid.appauth.AuthorizationService$TokenRequestTask.doInBackground(AuthorizationService.java:453)
        at net.openid.appauth.AuthorizationService$TokenRequestTask.doInBackground(AuthorizationService.java:395)
        at android.os.AsyncTask$3.call(AsyncTask.java:394)
```

This is Android stuff, just to report. 

- iOS: Bad redirection ❌ 

1. Regular request is sent -> returned 401 as unauthorized
2. Sent request to `oauth2/api/v1/token ` -> returns 302 pointing to `https://oc-10-5-0.oidc-1-0-0rc5.jw-qa.owncloud.works/login` -> returns 200
3. Check `status.php` -> 200
4. Check `.well-known/openid-configuration` -> 200

Then, login view with the Kopano option is visible in browser. After entering credentials and authorization correctly, browser do not hand over to app and shows the list of files ❌ 

Callback redirection is lost somewhere in the iOS re-sign in process. 

On the other hand, adding a new account from scratch after all these steps, works fine in iOS ✅ 

Android will need a new issue, because crashes must not happen whatever the reason. iOS, i could not find where the problem happened (client or server).

Devices:

Android: Google Pixel 2, Android v11
iOS: iPhoneXR v13.7

**Comment by jnweiger** [comment]
I am repeating steps 1), 3) and 4) with testpilotcloud 2.7.0daily20201019 (build 2311) Oct 15 2020 00:08:32

logging in, and logging out the client in each stage.
All logins succeed, first oauth2 is used, then later openidconnect is used. 
No problems :heart: 

Only when both, oauth2 and openidconnect are disabled, we see a slight glitch with https://github.com/owncloud/client/issues/8180

Step 2) is ommited, as in the new setup, the well-known url is automatically made visible and invisibe by enabling and disabling the openidconnect app.

**Comment by jesmrec** [comment]
Let's give a new try with `https://oc-10-5-0.oidc-1-0-0rc5-20201020.jw-qa.owncloud.works`

### Android

**1. OAuth2 enabled, OIDC disabled**

Connection correct, no issues ✅ 

**2. Enable OIDC, so both OAuth2 and OIDC are available**

- App uses OAuth2 and keeps using OAuth2  ✅ 
- By entering in the Account Manager -> Edit  account, the authentication process works correctly and the OAuth2 account becomes OIDC account ✅ 

**3. Disable OAuth2, so, only OIDC available**

Crash mentioned above happens ❌ 

No migration (install app from scratch and add account): works fine ✅ 

### iOS

**1. OAuth2 enabled, OIDC disabled**

Connection correct, no issues ✅ 

**2. Enable OIDC, so both OAuth2 and OIDC are available**

- App uses OAuth2 and keeps using OAuth2 ✅ 
- By entering in the Account Manager -> Edit  account, the authentication process works correctly but the list of files is shown in the browser ❌ 

**3. Disable OAuth2, so, only OIDC available**

List of files shown in the browser, not in the app ❌ 

No migration (install app from scratch and add account): works fine ✅

**Comment by pmaier1** [comment]
Docs PR https://github.com/owncloud/docs/pull/2830

**Comment by felix-schwarz** [comment]
Regarding triggering an upgrade of authentication methods by
- disabling authentication methods or 
- depending on letting authentication fail
presents its own challenges.

In that context, I proposed an alternative solution in (https://github.com/owncloud/ios-app/issues/824#issuecomment-718554614), but now reading this issue feel it really belongs into this issue:

Disabling authentication methods on the server to force the client to switch authentication methods could lock out client software that only supports Basic Auth and/or OAuth2.

The best solution IMO would be to provide information through the capabilities on migration paths for authentication methods, i.e. 

```json
"auth-migrations" : {
    "basic-auth" : [ "oidc", "oauth2" ],
    "oauth2" : "oidc"
}
```

This would indicate to clients that:
- if they use `basic-auth`, they should migrate users to `oidc` or `oauth2` (with `oidc` being preferred and `oauth2` provided as a fallback for clients not supporting `oidc`)
- if they use `oauth2`, they should migrate users to `oidc` if the client supports it
